README
NAME
    Test::MockFile - Allows tests to validate code that can interact with
    files without touching the file system.

VERSION
    Version 0.018

SYNOPSIS
    Intercepts file system calls for specific files so unit testing can take
    place without any files being altered on disk.

    This is useful for small tests
    <https://en.wikipedia.org/wiki/Google_Test#Small_Tests_(Unit_Tests)>
    where file interaction is discouraged.

    A strict mode is even provided which can throw a die when files are
    accessed during your tests!

        # Loaded before Test::MockFile so uses the core perl functions without any hooks.
        use Module::I::Dont::Want::To::Alter;

        use Test::MockFile;

        my $mock_file = Test::MockFile->file("/foo/bar", "contents\ngo\nhere");
        open(my $fh, "<", "/foo/bar") or die; # Does not actually open the file on disk.
        say "ok" if -e $fh;
        close $fh;
        say "ok" if (-f "/foo/bar");
        say "/foo/bar is THIS BIG: " . -s "/foo/bar"

        my $missing_mocked_file = Test::MockFile->file("/foo/baz"); # File starts out missing.
        my $opened = open(my $baz_fh, "<", "/foo/baz"); # File reports as missing so fails.
        say "ok" if !-e "/foo/baz";
    
        open($baz_fh, ">", "/foo/baz") or die; # open for writing
        print <$baz_fh> "replace contents\n";
    
        open($baz_fh, ">>", "/foo/baz") or die; # open for append.
        print <$baz_fh> "second line";
        close $baz_fh;
    
        say $baz->contents;
    
        # Unmock your file.
        undef $missing_mocked_file;
    
        # The file check will now happen on file system now the file is no longer mocked.
        say "ok" if !-e "/foo/baz";

IMPORT
    If the module is loaded in strict mode, any file checks, open, sysopen,
    opendir, stat, or lstat will throw a die.

    For example:

        use Test::MockFile qw/strict/;

        # This will not die.
        Test::MockFile->file("/bar", "...");
        Test::MockFile->symlink("/foo", "/bar");
        -l "/foo" or print "ok\n";
        open(my $fh, ">", "/foo");
    
        # All of these will die
        open(my $fh, ">", "/unmocked/file"); # Dies
        sysopen(my $fh, "/other/file", O_RDONLY);
        opendir(my $fh, "/dir");
        -e "/file";
        -l "/file"

SUBROUTINES/METHODS
  file
    Args: ($file, $contents, $stats)

    This will make cause $file to be mocked in all file checks, opens, etc.

    undef contents means that the file should act like it's not there.

    See "Mock Stats" for what goes in this hash ref.

  file_from_disk
    Args: "($file_to_mock, $file_on_disk, $stats)"

    This will make cause $file to be mocked in all file checks, opens, etc.

    If "file_on_disk" isn't present, then this will die.

    See "Mock Stats" for what goes in this hash ref.

  symlink
    Args: ($readlink, $file )

    This will cause $file to be mocked in all file checks, opens, etc.

    $readlink indicates what "fake" file it points to. If the file $readlink
    points to is not mocked, it will act like a broken link, regardless of
    what's on disk.

    If $readlink is undef, then the symlink is mocked but not present.(lstat
    $file is empty.)

    Stats are not able to be specified on instantiation but can in theory be
    altered after the object is created. People don't normally mess with the
    permissions on a symlink.

  dir
    Args: ($dir, \@contents, $stats)

    This will cause $dir to be mocked in all file checks, and opendir
    interactions.

    @contents should be provided in the sort order you expect to see the
    files from readdir. NOTE: Because "." and ".." will always be the first
    things readdir returns, These files are automatically inserted at the
    front of the array.

    See "Mock Stats" for what goes in this hash ref.

  Mock Stats
    When creating mocked files or directories, we default their stats to:

        Test::MockFile->file( $file, $contents, {
                'dev'       => 0,        # stat[0]
                'inode'     => 0,        # stat[1]
                'mode'      => $mode,    # stat[2]
                'nlink'     => 0,        # stat[3]
                'uid'       => 0,        # stat[4]
                'gid'       => 0,        # stat[5]
                'rdev'      => 0,        # stat[6]
                'atime'     => $now,     # stat[8]
                'mtime'     => $now,     # stat[9]
                'ctime'     => $now,     # stat[10]
                'blksize'   => 4096,     # stat[11]
                'fileno'    => undef,    # fileno()
        } );

    You'll notice that mode, size, and blocks have been left out of this.
    Mode is set to 666 (for files) or 777 (for directories), xored against
    the current umask. Size and blocks are calculated based on the size of
    'contents' a.k.a. the fake file.

    When you want to override one of the defaults, all you need to do is
    specify that when you declare the file or directory. The rest will
    continue to default.

        Test::MockFile->file("/root/abc", "...", {inode => 65, uid => 123, mtime => int((2000-1970) * 365.25 * 24 * 60 * 60 }));

        Test::MockFile->dir("/sbin", "...", { mode => 0700 }));

  new
    This class method is called by file/symlink/dir. There is no good reason
    to call this directly.

  contents
    Optional Arg: $contents

    Reports or updates the current contents of the file.

    To update, pass an array ref of strings for a dir or a string for a
    file. Symlinks have no contents.

  unlink
    Makes the virtual file go away. NOTE: This also works for directories.

  touch
    Optional Args: ($epoch_time)

    This function acts like the UNIX utility touch. It sets atime, mtime,
    ctime to $epoch_time.

    If no arguments are passed, $epoch_time is set to time(). If the file
    does not exist, contents are set to an empty string.

  stat
    Returns the stat of a mocked file (does not follow symlinks.)

  readlink
    Optional Arg: $readlink

    Returns the stat of a mocked file (does not follow symlinks.) You can
    also use this to change what your symlink is pointing to.

  is_link
    returns true/false, depending on whether this object is a symlink.

  is_dir
    returns true/false, depending on whether this object is a directory.

  is_file
    returns true/false, depending on whether this object is a regular file.

  size
    returns the size of the file based on its contents.

  exists
    returns true or false based on if the file exists right now.

  blocks
    Calculates the block count of the file based on its size.

  chmod
    Optional Arg: $perms

    Allows you to alter the permissions of a file. This only allows you to
    change the 07777 bits of the file permissions. The number passed should
    be the octal 0755 form, not the alphabetic "755" form

  permissions
    Returns the permissions of the file.

  mtime
    Optional Arg: $new_epoch_time

    Returns and optionally sets the mtime of the file if passed as an
    integer.

  ctime
    Optional Arg: $new_epoch_time

    Returns and optionally sets the ctime of the file if passed as an
    integer.

  atime
    Optional Arg: $new_epoch_time

    Returns and optionally sets the atime of the file if passed as an
    integer.

  add_file_access_hook
    Args: ( $code_ref )

    You can use add_file_access_hook to add a code ref that gets called
    every time a real file (not mocked) operation happens. We use this for
    strict mode to die if we detect your program is unexpectedly accessing
    files. You are welcome to use it for whatever you like.

    Whenever the code ref is called, we pass 2 arguments:
    "$code->($access_type, $at_under_ref)". Be aware that altering the
    variables in $at_under_ref will affect the variables passed to open /
    sysopen, etc.

    One use might be:

        Test::MockFile::add_file_access_hook(sub { my $type = shift; print "$type called at: " . Carp::longmess() } );

  clear_file_access_hooks
    Calling this subroutine will clear everything that was passed to
    add_file_access_hook

  How this mocking is done:
    Test::MockFile uses 2 methods to mock file access:

   -X via Overload::FileCheck
    It is currently not possible in pure perl to override stat
    <http://perldoc.perl.org/functions/stat.html>, lstat
    <http://perldoc.perl.org/functions/lstat.html> and -X operators
    <http://perldoc.perl.org/functions/-X.html>. In conjunction with this
    module, we've developed Overload::FileCheck.

    This enables us to intercept calls to stat, lstat and -X operators (like
    -e, -f, -d, -s, etc.) and pass them to our control. If the file is
    currently being mocked, we return the stat (or lstat) information on the
    file to be used to determine the answer to whatever check was made. This
    even works for things like "-e _". If we do not control the file in
    question, we return "FALLBACK_TO_REAL_OP()" which then makes a normal
    check.

   CORE::GLOBAL:: overrides
    Since 5.10, it has been possible to override function calls by defining
    them. like:

        *CORE::GLOBAL::open = sub(*;$@) {...}

    Any code which is loaded AFTER this happens will use the alternate open.
    This means you can place your "use Test::MockFile" statement after
    statements you don't want to be mocked and there is no risk that the
    code will ever be altered by Test::MockFile.

    We oveload the following statements and then return tied handles to
    enable the rest of the IO functions to work properly. Only open /
    sysopen are needed to address file operations. However opendir file
    handles were never setup for tie so we have to override all of opendir's
    related functions.

    *   open

    *   sysopen

    *   opendir

    *   readdir

    *   telldir

    *   seekdir

    *   rewinddir

    *   closedir

AUTHOR
    Todd Rinaldo, "<toddr at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    <https://github.com/CpanelInc/Test-MockFile>.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Test::MockFile

    You can also look for information at:

    *   CPAN Ratings

        <https://cpanratings.perl.org/d/Test-MockFile>

    *   Search CPAN

        <https://metacpan.org/release/Test-MockFile>

ACKNOWLEDGEMENTS
    Thanks to Nicolas R., "<atoomic at cpan.org>" for help with
    Overload::FileCheck. This module could not have been completed without
    it.

LICENSE AND COPYRIGHT
    Copyright 2018 cPanel L.L.C.

    All rights reserved.

    <http://cpanel.net>

    This is free software; you can redistribute it and/or modify it under
    the same terms as Perl itself. See perlartistic.



Changes
Revision history for Test-MockFile

0.033
- Request last Overload::FileCheck release - 0.013

0.032   2/24/2022
- Less strict mode - Don't complain about commands not opening or directly interacting with a file.

0.031   2/24/2022
- Fix for Plugin::FileTemp when calling tempfile in scalar context

0.030   2/22/2022
- Simplify _strict_mode_violation
- Introduce new_dir() to allow older dir() syntax.
- Provide a mechanism to setup custom strict rules
- Make is_strict_mode a helper
- Do not call _get_file_object when all we're trying to determine is if the file is mocked or not.
- Add Test::MockFile::Plugins
- We support "undef" as filename in strict rules to ignore filenames
- Block Plugin::FileTemp < 5.28
- Update MANIFEST

0.029   2/16/2022
- Fix broken link to small test documentation
- Ensure strict mode is enabled by default and prevent import conflicts.
- Fix warnings during global destruction
- Additional rules for _abs_path_to_file
- GH #103 - Update eg/example.pl:
- Add rmdir, mkdir and unlink default position
- Postpone file arg check during strict mode analysis
- Fixup hook args and teach for mkdir filename pos
- Perltidy policy changes.
- Use carp in filename and readlink
- Fix broken link to small test documentation

0.028   2/12/2022
- Update test for our FreeBSD results
- GH #122 - Support two-arg open, read/write, and understand pipes
- Add strict guards for additional keywords: readdir, telldir, rewinddir,
  seekdir, closedir, readlink, mkdir, rmdir
- Revert to old symbol resolution technique.

0.027   2/10/2022
- GH #75 - Strict mode is now on by default
- GH #45 - Provide a helpful error message and document fileno unsupported.
- GH #90 - Normalize forward slashes.
- GH #126 - Document using debugger under strict mode.
- Do not hard-code values of $!
- Try to get more data when t/runtime-bareword-filehandles.t fails on cpan testers.
- GH #64 -  Do not overwrite the error code when checking for Docker
- GH #63 -  FreeBSD's readlink() returns EINVAL for readlink(undef):
- Add more data for debugging test failures on OpenBSD

0.026   2/3/2022
- Fix support for mocking the top-level directory.
- Symlinks should appear in readdir
- Fix directory instantiation when creating a symlink.
- GH #105: Show directories in readdir
- Improve relative path management
- Don't let stat() get confused with trailing forward slash
- GH #118: Do not get confused by inner directory files.
- GH #85: Do not corrupt blessed handles.

0.025   1/26/2022
- Fix typo in chmod mock. Was accidentally calling chown.
- Fix dependency on Text::Glob. It is now a runtime requirement.

0.024   1/24/2022
- Prefer Carp::confess to die
- Do not use "$!" in tests as it's not consistent across platforms
- GH #78: Do not allow rmdir on a populated directory.
- GH #73: Prefer the term ->path to ->filename as it is more clear between file/dir
- chown $fh now works.
- Only warn about mixed files when using mocked files

0.024_01
 - GH #83: Get Solaris testers passing.
 - Remove all use of "$!" in tests as this has tranlation problems.

0.023   1/14/2022
 - GH #58: Fix synopsis typos
 - GH #65: Fix typos in the typo fixes.
 - GH #34: Support open() with barewords
 - GH #59: Detect and reject common path mistakes when mocking
 - GH #69: Redesigned dir() (and some file()) interface <--- breaks previous interface
 - GH #40: Support glob()
 - GH #15: Implement chown/chmod

0.022   12/27/2021
 -  GH #47: Manage bareword filehandles in runtime:

0.021   1/30/2020
 - Emit ENOTDIR on opendir when appropriate
 - Switch to github actions for CI testing

0.020   10/14/2019
 - GH #51 - Basic introduction of file ownership. Set default uid/gid to current user when not set
 - GH #49 - Make sysseek return position when called while allowing seek to return a true value.

0.19   5/21/2019
 - Fix POD for stat example in POD.
 - Allow scalar file handles on open. We don't care about those
 - Provide better guidance about keeping your mocks in scope in examples
 - Disable Test::CheckManifest due to break in Test-CheckManifest
 - Fix seek bug in sysopen(... O_APPEND)

0.018   1/24/2019
 - Properly handle open and sysopen file handles going out of scope.
 - Provide a helper method to mock objects to determine the file name. GH #31
 - Do not throw a file access hook when statting a file handle that is not under MockFile control. GH #30
 - Read multiple lines via readline when wantarray is true. GH #29

0.017   1/2/2019
 - Require a newer Test2::Bundle::Extended (0.000084) to support named isa_ok

0.016   1/2/2019
 - Bump Overload::FileCheck to 0.007 to address stack bug
 - Try to get EISDIR handling for BSD working. Apparently something changed in perl 5.20?
 - Do not do access hooks on filehandle interactions.

0.015   12/21/2018
 - Remove errant debug messages from open
 - Pass 3 to fix t/touch.t on BSD.

0.014   12/20/2018
 - Re-factor _find_file_or_fh to handle symlink following logic better. GH #26
 - Make more standard use of _get_file_object when looking up a file path so we
   properly handle abs path and symlink following

0.013   12/20/2018
 - Add a helper to determine if goto can be used. Where it's available is complicated
 - First pass at bareword file handles for opendir and friends
 - Fix errant docs for making a symlink mock
 - Follow links for stat but not lstat
 - Add support for readlink
 - Try 2 to fix BSD issues with GH #20

0.012   11/16/2018
 - Fix for #21 - length undef on perl 5.10
 - GH #20 - Add EPERM support for freebsd when unlinking directories
 - Fix for print $fh undef throwing a warning
 - Remove faulty OS level test of readdir after opendir.

0.011   11/08/2018
 - Fix for my $file_contents = do { local $/; <$fh> };
 - Mock CORE::GLOBAL::unlink and support directories for unlink
 - Mock CORE::GLOBAL::mkdir
 - Mock CORE::GLOBAL::rmdir
 - MockFile->symlink now follows the symlink norm of ($target, $file), not the reverse.
 - New mock helper 'exists' to check if the file is there.

lib/Test/MockFileSys.pm
package Test::MockFileSys;

use Test::MockFile ();

my $singleton;

sub _automock_hook;

sub import {
    my ( $class, @args ) = @_;

    foreach my $arg (@args) {
        Test::MockFile->import($arg) if ( $arg =~ m/^(?:no)strict$/ );
    }

    Test::MockFile::set_strict_mode_automock( \&_automock_hook );

    return;
}

sub is_strict_mode {
    return $Test::MockFile::STRICT_MODE_STATUS == Test::MockFile::STRICT_MODE_DISABLED() ? 0 : 1;
}

sub _automock_hook {
    my ($path) = @_;
    return unless ref $singleton;

    my $mock = Test::MockFile->new( { 'path' => $path, 'contents' => undef } );
    return $singleton->{'files'}->{ $mock->path } = $mock;
}

# Grabs existing objects?
#my $fs = Test::MockFileSys->new; # Singleton Only one at a time.

sub new {
    my ( $class, @args ) = @_;

    !defined $singleton or die("$class has already been instantiated");

    $singleton = bless {}, $class;
    Scalar::Util::weaken($singleton);

    $singleton->init;

    return $singleton;
}

sub init {
    my ($self) = @_;

    $self->{'files'} = {};

    return;
}

sub get_mocked_path {
    my ( $self, $path, $type ) = @_;
    my $abs_path = Test::MockFile::_abs_path_to_file($path);
    my $m        = $self->{'files'}->{$abs_path};
    return $m if $m;    # We hold the object. Just return it.

    # Did someone directly mock it in T::MF? If so we're going to store it here also.
    $m = $Test::MockFile::files_being_mocked{$abs_path};
    if ($m) {
        $self->{'files'}->{$abs_path} = $m;
        return $m;
    }

    return unless defined $type && length $type;

    # We're going to have to instantiate it and we will own it.
    return $self->{'files'}->{$abs_path} = Test::MockFile->can($type)->( 'Test::MockFile', $abs_path );
}

## Just mock not there
#$fs->dir('/foo');
#$fs->file('/foo/bar');
#$fs->symlink('/foo/bar');

sub dir {
    my ( $self, $path ) = @_;

    return $self->get_mocked_path( $path, 'dir' );
}

sub file {
    my ( $self, $path ) = @_;

    return $self->get_mocked_path( $path, 'file' );
}

sub symlink {
    my ( $self, $path ) = @_;

    return $self->get_mocked_path( $path, 'symlink' );
}

# Get the object
#$fs->path('/foo/bar');

sub path {
    my ( $self, $path ) = @_;
    return $self->get_mocked_path($path);
}

## unmock the thing.
# NOTE: it will not unmock if something is also holding the object.
#$fs->unmock('/foo/bar');
sub unmock {
    my ( $self, $path ) = @_;
    my $mock = $self->path($path);

    delete $self->{'files'}->{ $mock->path };
}

## make the thing and set perms, etc.
#$fs->write_file('/foo/bar', "", { perms => 0755 });
#$fs->symlink('data', '/foo/link');
#$fs->mkdir( '/foo', 0755); # Auto mocks in strict mode.

sub write_file {
    my ( $self, $path, $content, $stats );

    my $mock = $self->file($path);
    $mock->{'content'} = $content if defined $content;

    #$mock->set_stats($stats) if ref $stats;

    return $mock;
}
__END__
sub file_contents {
    my ($self, $path) = @_;
    
    my $mock = $self->file($path);
    return $mock->{'content'};
}

## This is broken right now.
#sub symlink {
#    my ($self, $path, $content, $stats );
#    
#    my $mock = $self->symlink($path);
#    $mock->{'readlink'} = $content if defined $content;
#    $mock->set_stats($stats) if ref $stats;
#    
#    return $mock;
#}

sub mkdir {
    my ($self, $path, $perms, $stats );
    
    my $mock = $self->dir($path);
    $mock->chmod($perms) if length $perms;    
    #$mock->set_stats($stats) if ref $stats;
    
    return $mock;
}


## automock in strict mode.
#symlink('data', '/foo/link');
#mkdir '/foo', 0755;
#chmod 0755, '/foo/bar';
#unlink '/foo/link';
#
## Get the object


1;


lib/Test/MockFile/Stat.pm
# Copyright (c) 2018, cPanel, LLC.
# All rights reserved.
# http://cpanel.net
#
# This is free software; you can redistribute it and/or modify it under the
# same terms as Perl itself. See L<perlartistic>.

package Test::MockFile::Stat;

use strict;
use warnings;

# From http://man7.org/linux/man-pages/man7/inode.7.html
use constant S_IFMT    => 0170000;    # bit mask for the file type bit field
use constant S_IFPERMS => 07777;      # bit mask for file perms.

use constant S_IFSOCK => 0140000;     # socket
use constant S_IFLNK  => 0120000;     # symbolic link
use constant S_IFREG  => 0100000;     # regular file
use constant S_IFBLK  => 0060000;     # block device
use constant S_IFDIR  => 0040000;     # directory
use constant S_IFCHR  => 0020000;     # character device
use constant S_IFIFO  => 0010000;     # FIFO

sub new {
    my $class = shift @_;

    my %opts;
    if ( scalar @_ == 1 ) {
        %opts = %{ $_[0] };
    }
    else {
        %opts = @_;
    }

    my $now = time;

    my $self = bless {
        'dev'      => 0,        # stat[0]
        'inode'    => 0,        # stat[1]
        'mode'     => 0,        # stat[2]
        'nlink'    => 0,        # stat[3]
        'uid'      => 0,        # stat[4]
        'gid'      => 0,        # stat[5]
        'rdev'     => 0,        # stat[6]
        'size'     => undef,    # stat[7]
        'atime'    => $now,     # stat[8]
        'mtime'    => $now,     # stat[9]
        'ctime'    => $now,     # stat[10]
        'blksize'  => 4096,     # stat[11]
        'blocks'   => 0,        # stat[12]
        'fileno'   => undef,    # fileno()
        'tty'      => 0,        # possibly this is already provided in mode?
        'readlink' => '',       # what the symlink points to.
    }, $class;

    foreach my $key ( keys %opts ) {

        # Ignore Stuff that's not a valid key for this class.
        next unless exists $self->{$key};

        # If it's passed in, we override them.
        $self->{$key} = $opts{$key};
    }

    $self->{'fileno'} //= _unused_fileno();

    if ( $self->{'size'} && !$self->{'blocks'} ) {
        $self->resize;
    }

    return $self;
}

sub get_stats {
    my $self = shift;

    return (
        $self->{'dev'},        # stat[0]
        $self->{'inode'},      # stat[1]
        $self->{'mode'},       # stat[2]
        $self->{'nlink'},      # stat[3]
        $self->{'uid'},        # stat[4]
        $self->{'gid'},        # stat[5]
        $self->{'rdev'},       # stat[6]
        $self->{'size'},       # stat[7]
        $self->{'atime'},      # stat[8]
        $self->{'mtime'},      # stat[9]
        $self->{'ctime'},      # stat[10]
        $self->{'blksize'},    # stat[11]
        $self->{'blocks'},     # stat[12]
    );
}

sub _unused_fileno {
    return 900;                # TODO
}

# Helpers for making file/link/dir
sub file {
    my ( $class, %opt ) = @_;

    my $perms = defined $opt{'mode'} ? int( $opt{'mode'} ) : 0666;
    $opt{'mode'} = ( $perms ^ umask ) & S_IFREG;

    return $class->new( \%opt );
}

sub dir {
    my ( $class, %opt ) = @_;

    my $perms = defined $opt{'mode'} ? int( $opt{'mode'} ) : 0777;
    $opt{'mode'} = ( $perms ^ umask ) & S_IFDIR;

    return $class->new( \%opt );
}

sub link {
    my ( $class, $target_file, %opt ) = @_;

    length $target_file or die("Symlinks must point to a file.");

    $opt{'readlink'} = $target_file;
    $opt{'mode'}     = 0777 | S_IFLNK;

    return $class->new( \%opt );
}

sub readlink {
    my ($self) = @_;

    return $self->{'readlink'};
}

sub is_link {
    my ($self) = @_;

    return ( length $self->{'readlink'} && $self->{'mode'} & S_IFLNK ) ? 1 : 0;
}

sub resize {
    my ( $self, $new_size ) = @_;

    $self->{'size'} = abs( int( $new_size || 0 ) );
    $self->{'blocks'} = $self->{'size'} / abs( $self->{'blksize'} || 1 );
}

sub chmod {
    my ( $self, $mode ) = @_;

    $mode = int($mode) | S_IFPERMS;

    $self->{'mode'} = ( $self->{'mode'} & S_IFMT ) + $mode;

    return $mode;
}

sub mtime {
    my ( $self, $time ) = @_;

    $time = time unless defined $time && $time =~ m/^[0-9]+$/;
    $self->{'mtime'} = $time;
}

sub ctime {
    my ( $self, $time ) = @_;

    $time = time unless defined $time && $time =~ m/^[0-9]+$/;
    $self->{'ctime'} = $time;
}

sub atime {
    my ( $self, $time ) = @_;

    $time = time unless defined $time && $time =~ m/^[0-9]+$/;
    $self->{'atime'} = $time;
}

1;


Makefile.PL
use 5.010;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Test::MockFile',
    AUTHOR             => q{Todd Rinaldo <toddr@cpanel.net>},
    VERSION_FROM       => 'lib/Test/MockFile.pm',
    ABSTRACT_FROM      => 'lib/Test/MockFile.pm',
    LICENSE            => 'artistic_2',
    PL_FILES           => {},
    MIN_PERL_VERSION   => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More'                => '1.302133',
        'Test2::Bundle::Extended'   => '0.000084',    # Oldest version provided on CPAN isa_ok changed in 0.000035 and we need that.
        'Test2::Tools::Explain'     => '0',
        'Test2::Plugin::NoWarnings' => '0',
        'File::Slurper'             => 0,
        'Text::Glob'                => 0,
    },
    PREREQ_PM => {
        'Overload::FileCheck' => '0.012',
    },
    dist       => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean      => { FILES    => 'Test-MockFile-*' },
    META_MERGE => {
        resources => {
            license    => 'http://dev.perl.org/licenses/',
            bugtracker => 'https://github.com/cpanelinc/Test-MockFile/issues',
            repository => 'https://github.com/cpanelinc/Test-MockFile',
        }
    },
);
